src/iterators/motivation.md

```diff
@@ -0,0 +1,48 @@
+---
+minutes: 3
+---
+
+# Motivating Iterators
+
+If you want to iterate over the contents of an array, you'll need to define:
+
+- Some state to keep track of where you are in the iteration process, e.g. an
+  index.
+- A condition to determine when iteration is done.
+- Logic for updating the state of iteration each loop.
+- Logic for fetching each element using that iteration state.
+
+In a C-style for loop you declare these things directly:
+
+```c,editable
+for (int i = 0; i < array_len; i += 1) {
+    int elem = array[i];
+}
```

Comment:
As I put this PR up it occurs to me that we don't need to use C-style `for` to demonstrate this concept, we could show the same thing in pure Rust:
```rust
let mut i = 0;
while i < array.len() {
    let elem = array[i];
    i += 1;
}
```
Might still be good to mention C-style `for` in the speaker notes, since it's still a common point of reference.

Reply:
I kinda still feel like the C version is going to be clearer and more familiar for most students, tho

Reply:
@djmitche Do you have an opinion here? I'm leaning towards leaving this as a c-style `for` for readability and familiarity

Reply:
You know what, I'm going to leave it as a C-style `for` but I'm going to add the rust equivalent in the speaker notes for easy reference. Let me know if you think that's a bad idea for whatever reason, though.

Reply:
I think the C-style is fine.